File: src/test/java/aww/yeah/TwoHundredControllerIntegrationTest.java
package aww.yeah;

import com.jayway.restassured.RestAssured;
import com.jayway.restassured.response.Response;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.core.io.InputStreamResource;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;

import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static com.jayway.restassured.RestAssured.given;
import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class TwoHundredControllerIntegrationTest {

    private final static char[] CHARS = IntStream.rangeClosed('A', 'Z')
            .mapToObj(c ->  "" + (char)(c + 32) + (char) c).collect(Collectors.joining()).toCharArray();

    @LocalServerPort
    private int port;

    private final Random random = new Random();

    @BeforeEach
    public void setUp() throws Exception {
        RestAssured.port = port;
    }

    @Test
    public void twoHundredApplication_shouldRespondToRandomGetRequests_withStatusOk() throws Exception {
        for(int i=0; i<20;i++) {

            Response response = given()
                    .get(genRandomString())
                    .thenReturn();

            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK.value());

            if(MediaType.APPLICATION_JSON_VALUE.equals(response.getContentType())) {
                assertThat(response.getBody().asString()).isEqualTo("{\"success\":\"Aww Yeah\"}");
            } else {
                assertThat(response.getBody().asInputStream())
                        .hasSameContentAs(new InputStreamResource(ClassLoader.getSystemResourceAsStream("aww_yeah.gif")).getInputStream());
            }
        }
    }

    @Test
    public void twoHundredApplication_shouldRespondToRandomGetRequestsWithUriVars_withStatusOk() throws Exception {

        for(int i=0; i<20;i++) {
            Response response = given()
                    .param(genRandomString(), genRandomString())
                    .param(genRandomString(), genRandomString())
                    .get(genRandomString())
                    .thenReturn();

            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK.value());
            if(MediaType.APPLICATION_JSON_VALUE.equals(response.getContentType())) {
                assertThat(response.getBody().asString()).isEqualTo("{\"success\":\"Aww Yeah\"}");
            } else {
                assertThat(response.getBody().asInputStream())
                        .hasSameContentAs(new InputStreamResource(ClassLoader.getSystemResourceAsStream("aww_yeah.gif")).getInputStream());
            }
        }
    }

    @Test
    public void twoHundredApplication_shouldRespondToRandomPostWithUriVarRequests_withStatusOk() throws Exception {

        for(int i=0; i<20;i++) {

            Response response = given()
                    .queryParam(genRandomString(), genRandomString())
                    .queryParam(genRandomString(), genRandomString())
                    .post(genRandomString())
                    .thenReturn();

            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK.value());
            if(MediaType.APPLICATION_JSON_VALUE.equals(response.getContentType())) {
                assertThat(response.getBody().asString()).isEqualTo("{\"success\":\"Aww Yeah\"}");
            } else {
                assertThat(response.getBody().asInputStream())
                        .hasSameContentAs(new InputStreamResource(ClassLoader.getSystemResourceAsStream("aww_yeah.gif")).getInputStream());
            }
        }
    }

    @Test
    public void twoHundredApplication_shouldRespondToRandomPostWithFormDataRequests_withStatusOk() throws Exception {

        for(int i=0; i<20;i++) {

            Response response = given()
                    .formParam(genRandomString(), genRandomString())
                    .formParam(genRandomString(), genRandomString())
                    .post(genRandomString())
                    .thenReturn();

            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK.value());
            if(MediaType.APPLICATION_JSON_VALUE.equals(response.getContentType())) {
                assertThat(response.getBody().asString()).isEqualTo("{\"success\":\"Aww Yeah\"}");
            } else {
                assertThat(response.getBody().asInputStream())
                        .hasSameContentAs(new InputStreamResource(ClassLoader.getSystemResourceAsStream("aww_yeah.gif")).getInputStream());
            }
        }
    }

    private String genRandomString() {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < random.nextInt( 15) + 5; i++) {
            builder.append(CHARS[random.nextInt(CHARS.length)]);
        }
        return builder.toString();
    }
}

File: src/test/kotlin/aww/yeah/TwoHundredControllerIntegrationTest.kt
package aww.yeah

import com.jayway.restassured.RestAssured
import com.jayway.restassured.http.ContentType
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.core.io.InputStreamResource
import org.springframework.http.HttpStatus
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import kotlin.random.Random


@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
internal class TwoHundredControllerIntegrationTest {
    @LocalServerPort
    private val port = 0

    @MockBean
    private lateinit var randomBean: Random

    @BeforeEach
    internal fun setUp() {
        RestAssured.port = port
    }

    private val awwYeahGif = ClassLoader.getSystemResourceAsStream("aww_yeah.gif")

    @Test
    internal fun `two hundred application should respond to random GET requests with status 200 Ok`() {
        for (i in 1..4) {
            `when`(randomBean.nextInt(5)).thenReturn(i)

            val response = RestAssured.given()
                    .get(genRandomString())
                    .then()
                    .statusCode(HttpStatus.OK.value())
                    .and()
                    .contentType(ContentType.JSON)
                    .and()
                    .extract().response().asString()

            assertThat(response).isEqualTo("""{"success":"Aww Yeah"}""")
        }
    }

    @Test
    internal fun `two hundred application should respond to random GET requests with status 200 Ok and GIF if random is 0`() {
        `when`(randomBean.nextInt(5)).thenReturn(0)

        val response = RestAssured.given()
                .get(genRandomString())
                .then()
                .statusCode(HttpStatus.OK.value())
                .and()
                .contentType("image/gif")
                .and()
                .extract().response().asInputStream()

        assertThat(response)
                .hasSameContentAs(InputStreamResource(awwYeahGif).inputStream)
    }

    @Test
    internal fun `two hundred application should respond to random GET requests that have random query params with status 200 Ok`() {
       for(i in 1..4) {
           `when`(randomBean.nextInt(5)).thenReturn(i)

           val response = RestAssured.given()
                   .param(genRandomString(), genRandomString())
                   .param(genRandomString(), genRandomString())
                   .get(genRandomString())
                   .then()
                   .statusCode(HttpStatus.OK.value())
                   .and()
                   .contentType(ContentType.JSON)
                   .and()
                   .extract().response().asString()

           assertThat(response).isEqualTo("""{"success":"Aww Yeah"}""")
       }
    }


    @Test
    internal fun `two hundred application should respond to random GET requests that have random query params with status 200 Ok and GIF if random is 0`() {
        `when`(randomBean.nextInt(5)).thenReturn(0)

        val response = RestAssured.given()
                .queryParam(genRandomString(), genRandomString())
                .queryParam(genRandomString(), genRandomString())
                .post(genRandomString())
                .then()
                .statusCode(HttpStatus.OK.value())
                .and()
                .contentType("image/gif")
                .and()
                .extract().response().asInputStream()

        assertThat(response)
                .hasSameContentAs(InputStreamResource(awwYeahGif).inputStream)
    }

    @Test
    fun `two hundred application should respond to random POST requests that have random form params with status 200 Ok`() {
        for(i in 1..4) {
            `when`(randomBean.nextInt(5)).thenReturn(i)

            val response = RestAssured.given()
                    .formParam(genRandomString(), genRandomString())
                    .formParam(genRandomString(), genRandomString())
                    .post(genRandomString())
                    .then()
                    .statusCode(HttpStatus.OK.value())
                    .and()
                    .contentType(ContentType.JSON)
                    .and()
                    .extract().response().asString()

            assertThat(response).isEqualTo("""{"success":"Aww Yeah"}""")
        }
    }

    @Test
    internal fun `two hundred application should respond to random POST requests that have random form params with status 200 Ok and GIF if random is 0`() {
        `when`(randomBean.nextInt(5)).thenReturn(0)

        val response = RestAssured.given()
                .formParam(genRandomString(), genRandomString())
                .formParam(genRandomString(), genRandomString())
                .post(genRandomString())
                .then()
                .statusCode(HttpStatus.OK.value())
                .and()
                .contentType("image/gif")
                .and()
                .extract().response().asInputStream()

        assertThat(response)
                .hasSameContentAs(InputStreamResource(awwYeahGif).inputStream)
    }

    private fun genRandomString(): String =
            IntRange(0, Random.nextInt(15) + 5)
                    .map { asciiNonCtrlChars[Random.nextInt(asciiNonCtrlChars.size)] }
                    .joinToString(separator = "")
                    .let { URLEncoder.encode(it, StandardCharsets.UTF_8.toString()) }

    private val asciiNonCtrlChars =
            CharRange('!', '~')
                    .joinToString(separator = "")
                    .toCharArray()
}
